### PR TITLE
Source Treeview

### PR DESCRIPTION
**Pull request by matandobr**
! ONGOING PR !

- Tested on Windows til now
- Changed the Java view to the tree-view
- Used jsTree library for the javascript tree-view (@ajinabraham - do we need to add their license?)
- The whole data is generated within the HTML with a os.dir + template generator hack, big thanks to Elf Sternberg from [this thread ](https://stackoverflow.com/questions/6297068/whats-the-django-way-to-render-a-tree-of-folders-and-files)
- For now the filename filtering is done in the client side, and the content search is done both in the client and server side (Query the find source view for all the files that match the query -> get all of the files and only "higlight" them in the client side - actually I search the whole paths with a hack in the client side, refer to the new search callback for the implementation...)

### CODE REVIEW

File: StaticAnalyzer/views/android/java.py [hunk 1]
```diff
@@ -7,14 +7,32 @@
...
 logger = logging.getLogger(__name__)
 
 
+# Generator that uses 2 template files in order to make the main template
+def tree_index_maker(root_dir):
+    def _index(root):
+        files = os.listdir(root)
+        for mfile in files:
+            t = os.path.join(root, mfile)
+            if os.path.isdir(t):
```
[the remark this answers is not in the document]
Comment (matandobr) [reply StaticAnalyzer/views/android/java.py]: I didn't get why should we argue with the python cross platform solution with something that is simple as directory checking..  isn't is complicating a simple thing?

[the remark this answers is not in the document]
Comment (ajinabraham) [reply StaticAnalyzer/views/android/java.py]: So reason for this is that I was thinking it's nice to replace os.path with the new pathlib
https://docs.python.org/3/library/pathlib.html If we stick with a function from utils, it is so easy to maintain and update this.

[the remark this answers is not in the document]
Comment (matandobr) [reply StaticAnalyzer/views/android/java.py]: Got you,
I'll stick with the projects standard of course.

**Comment by ajinabraham** [comment]
@matandobr Got a chance to look into the review comments?

**Comment by matandobr** [comment]
Hey @ajin sorry for the late reply.
I'll try to finish this weekend.

**Comment by matandobr** [comment]
Tested on Ubuntu 18 
image

**Comment by matandobr** [comment]
@ajinabraham @superpoussin22 
Hey, Ready for review and testing :)

**Comment by ajinabraham** [comment]
Gonna take a dig!

**Comment by matandobr** [comment]
@ajinabraham 
Also, a thought - does it makes sense to add a package / app name in the title? 
Like 
Java source - a.b.c v4.3.2

**Comment by ajinabraham** [comment]
Yes it does

**Comment by ajinabraham** [comment]
@matandobr I am still getting crashed/slow browser for a pretty large APK like Spotify on a couple of file name search. 
Tested on Chrome and Firefox.

Screenshot 2020-09-26 at 2 03 49 PM

Can search filename only on enter key improve this behavior?

And for file content search inside a large APK source. It's incredibly slow compared to server-side searching. Can we move this logic to server side AJAX calls to speed that up? It's either browser restrict CPU/RAM to tabs or client side JavaScript can't do this heavy lifting due to browser imposed restrictions or we have leaks somewhere.

I really like this single page UI especially the searching part and awesome UX, but we need to be stable and consistent across large/small files.
https://streamable.com/0sqzqp
On Macbook pro, Core i7 2.7GHz, 16GB RAM

**Comment by matandobr** [comment]
> @matandobr I am still getting crashed/slow browser for a pretty large APK like Spotify on a couple of file name search.
> Tested on Chrome and Firefox.
> 
> Screenshot 2020-09-26 at 2 03 49 PM
> 
> Can search filename only on enter key improve this behavior?
> 
> And for file content search inside a large APK source. It's incredibly slow compared to server-side searching. Can we move this logic to server side AJAX calls to speed that up? It's either browser restrict CPU/RAM to tabs or client side JavaScript can't do this heavy lifting due to browser imposed restrictions or we have leaks somewhere.
> 
> I really like this single page UI especially the searching part and awesome UX, but we need to be stable and consistent across large/small files.
> https://streamable.com/0sqzqp
> On Macbook pro, Core i7 2.7GHz, 16GB RAM

@ajinabraham , Are you actually experiencing a "crush"? - the video you sent is just loading, maybe it will actually appear after some time, also you searched for an extremely common word in the content search,
I 'm not able to see the picture you uploaded, could you send me on slack?

**Comment by ajinabraham** [comment]
@matandobr The page becomes unresponsive and the browser interferes after being unresponsive after a while. The results might appear after prolonged waiting, but that is quite a performance impact. The server-side implementation of the same is much faster. Can we change the logic to use server-side search?

**Comment by matandobr** [comment]
@ajinabraham 
I'll try to change the filename search to server-side.
Besides that, the content just really takes a lot of time, also in the python side. 
I looked at find.py and I think we could make this more efficient - 
image
I think reading the whole source files isn't the right way to go, If we are trying to search a text within those files I think reading line by line would be extremely more efficient.

**Comment by matandobr** [comment]
@ajinabraham After some debugging, the thing that takes most of the time is the searching within the the JS side,
I found a way to overcome it, on it.

**Comment by matandobr** [comment]
Changed the Filename to server side + JS side improvements

**Comment by matandobr** [comment]
@ajinabraham Fixed all latest rejects, 
Works awesome now.

File: StaticAnalyzer/views/android/find.py [hunk 2]
```diff
@@ -15,49 +18,67 @@
...
     try:
-        if not re.match('^[0-9a-f]{32}$', request.POST['md5']):
-            return print_n_send_error_response(request, 'Searching Failed')
+        match = re.match('^[0-9a-f]{32}$', request.POST['md5'])
+        if not match:
+            raise ValueError('Invalid MD5 hash')
         md5 = request.POST['md5']
         query = request.POST['q']
         code = request.POST['code']
-        matches = set()
```
Comment (ajinabraham) [inline StaticAnalyzer/views/android/find.py:25]: @matandobr This section added support for kotlin, and android zipped source code , both eclipse and android studio. This looks like removed in the PR

**Review by ajinabraham — CHANGES_REQUESTED** [review]
Android Source ZIP file view functions do not work, also kotlin support looks removed as per https://github.com/MobSF/Mobile-Security-Framework-MobSF/pull/1364/files#diff-5e1eb72d457ddf6666cb55be6f61c9b7L42

**Review by ajinabraham — COMMENTED** [review]
Code Looks Great @matandobr 
Testing this out.

File: StaticAnalyzer/views/android/find.py [hunk 3]
```diff
@@ -3,61 +3,70 @@
...
-            'md5': md5,
-            'typ': code,
-            'found': str(len(matches)),
+            'found': str(flz),
+            'search_type': search_type,
             'version': settings.MOBSF_VER,
         }
-        template = 'general/search.html'
-        return render(request, template, context)
+        return JsonResponse(json.dumps(context), safe=False)
```
[the remark this answers is not in the document]
Comment (matandobr) [reply StaticAnalyzer/views/android/find.py:69]: Inherited from previous code I guess? 
no particular reason..

[the remark this answers is not in the document]
Comment (matandobr) [reply StaticAnalyzer/views/android/find.py:69]: OK, I guess that this is the reason 
image

[the remark this answers is not in the document]
Comment (matandobr) [reply StaticAnalyzer/views/android/find.py:69]: This is due the list inside the dict... there is no easy way to overcome it

**Comment by ajinabraham** [comment]
@matandobr Both File and Content search does not return any results now.

**Comment by matandobr** [comment]
> @matandobr Both File and Content search does not return any results now.

🤔 Any errors?

**Comment by ajinabraham** [comment]
Something with the UI may be?

Screenshot 2020-10-13 at 4 29 06 PM

**Comment by ajinabraham** [comment]
@matandobr Tried this up on a fresh instance. Still the same. I think the search is happening but the UI is not updated accordingly.
Screen Shot 2020-10-18 at 4 43 46 PM

**Comment by matandobr** [comment]
@ajinabraham 
Fixed the search issue in mac/linux

**Comment by ajinabraham** [comment]
@matandobr I made some fixes to remove slash modifications which can be a pain in the ass across Windows/*nix.
Can you test it in windows? I don't have access to Windows until next month.
Also please remove this `replace()` if it is not relevant. https://github.com/MobSF/Mobile-Security-Framework-MobSF/pull/1364/files#diff-fd455d570d21d6d55a8fbc8883076ffe2a712c26d7cab5c79f9f644342342f0cR105

**Comment by matandobr** [comment]
Hey @ajinabraham thank you for the review, 
your modification for the sanitized slashes is exactly what I fixed in https://github.com/MobSF/Mobile-Security-Framework-MobSF/pull/1364/commits/5540bd82ed779dd333221d155f51748b94a47154

So no it does not work now...
image

**Comment by ajinabraham** [comment]
@matandobr That fix had issues. The find was successful but the file read was failing in *nix.

**Comment by matandobr** [comment]
@ajinabraham Are you sure? I tested it on linux 👍

**Comment by ajinabraham** [comment]
@matandobr  The find was successful but the file read was failing in *nix. 


The json matches on the screenshot you shared looks good. Why is the view source failing in windows?
Is the path joined properly in windows https://github.com/matandobr/Mobile-Security-Framework-MobSF/blob/JAVA_TREEVIEW/StaticAnalyzer/views/android/view_source.py#L54

**Comment by matandobr** [comment]
It is, but the reason it fails is that the search in the javascript is failing is because the separators in the HTML (due to the windows seperators) are \ and not /
So the search is returned great from the server but the select_nodes function is failing in the JS side,

**Comment by ajinabraham** [comment]
@matandobr Tested ok Linux as well.

So where should we fix this? I was using pathlib so that it gives a consistent path format irrespective of OS. It can handle both windows and nix path. Does the issue needs a fix at JS side?

**Comment by matandobr** [comment]
@ajinabraham 
I guess the right way to go is to change the tree_index_maker function within source_tree.py so it will use pathlib also... I'll try to fix it tomorrow

**Comment by ajinabraham** [comment]
With that fix, this PR is good to merge 👍

**Comment by matandobr** [comment]
great 👍

**Review by ajinabraham — APPROVED** [review]
LGTM 👍 Thanks for improving upon this.

**Comment by ajinabraham** [comment]
Closing in favour of https://github.com/MobSF/Mobile-Security-Framework-MobSF/pull/1572

**Pull request closed without merging**